### ISSUE TITLE
Segmentation fault in EVP_PKEY_Q_keygen()

### ISSUE THREAD

**Issue by sgallagher** [openssl/openssl#18546]
This is using OpenSSL from Fedora 36: `openssl-libs-3.0.3-1.fc36.x86_64`

I also reproduced it with `openssl-1:3.0.2-4.fc36.x86_64`

But if I use `openssl-libs-3.0.0-1.fc36.x86_64`, the problem does not appear. I assume it's a regression introduced somewhere between 3.0.0 and 3.0.2. I don't have an RPM of 3.0.1 around anywhere to test with.

Trivial reproducer, compiled with:
`gcc -o rsagen -O0 -ggdb -I/usr/include -lcrypto rsagen.c`  using gcc (GCC) 12.0.1 20220413 (Red Hat 12.0.1-0)

Also verified with clang version 14.0.0 (Fedora 14.0.0-1.fc36), same arguments.
```C
#include <openssl/evp.h>
#include <openssl/rsa.h>


int main(int argc, char **argv)
{
  EVP_PKEY *pkey = NULL;
  pkey = EVP_PKEY_Q_keygen(NULL, NULL, "RSA", 2048);
  if (!pkey)
  {
    return 1;
  }

  return 0;
}
```

Full backtrace:

```
(gdb) bt full
#0  __strcasecmp_l_avx () at ../sysdeps/x86_64/multiarch/strcmp-sse42.S:140
No locals.
#1  0x00007ffff7d1e7f6 in EVP_PKEY_Q_keygen (libctx=0x0, propq=0x0, type=0x402010 "RSA") at crypto/evp/evp_lib.c:1172
        args = {{gp_offset = 24, fp_offset = 48, overflow_arg_area = 0x7fffffffdef0, reg_save_area = 0x7fffffffde20}}
        bits = 0
        name = <optimized out>
        params = {{key = 0x0, data_type = 0, data = 0x0, data_size = 0, return_size = 0}, {key = 0x0, data_type = 0, data = 0x0, data_size = 0, return_size = 0}}
        ret = 0x0
        __func__ = "EVP_PKEY_Q_keygen"
#2  0x000000000040115b in main (argc=1, argv=0x7fffffffe028) at rsagen.c:8
        pkey = 0x0
```

**Comment by sgallagher** [comment]
I unrolled that trivial reproducer a bit, but I originally encountered it with the even simpler:

```C
#include <openssl/evp.h>
#include <openssl/rsa.h>

int main(int argc, char **argv)
{
  EVP_PKEY *pkey = NULL;
  pkey = EVP_RSA_gen(2048);
  if (!pkey)
  {
    return 1;
  }

  return 0;
}
```

**Comment by beldmit** [comment]
It should be fixed in current OpenSSL 3.0 branch and I will fix it in Fedora when 3.0.4 will be out.

**Comment by sgallagher** [comment]
OK, looks like it's addressed by #18247 and #18172 

Thanks!
